### PR TITLE
Fix/flake TestNewManagerImplStartProbeMode  unit test

### LINKED ISSUE #121451: [Flake] unit test TestNewManagerImplStartProbeMode
Seen in https://github.com/kubernetes/kubernetes/pull/121445#issuecomment-1775347291 and in https://github.com/kubernetes/kubernetes/pull/121407

No hits on https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=TestNewManagerImplStartProbeMode

stack strace

```
E1023 13:40:22.589536   43369 goroutinemap.go:150] Operation for "/tmp/device_plugin1067413082/server.sock" failed. No retries permitted until 2023-10-23 13:40:23.0894569 +0000 UTC m=+0.753948700 (durationBeforeRetry 500ms). Error: RegisterPlugin error -- failed to get plugin info using RPC GetInfo at socket /tmp/device_plugin1067413082/server.sock, err: rpc error: code = Unimplemented desc = unknown service pluginregistration.Registration
--- FAIL: TestNewManagerImplStartProbeMode (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x28fb0ac]

goroutine 238 [running]:
testing.tRunner.func1.2({0x2b14fa0, 0x452d660})
	/usr/local/go/src/testing/testing.go:1545 +0x366
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1548 +0x630
panic({0x2b14fa0?, 0x452d660?})
	/usr/local/go/src/runtime/panic.go:920 +0x270
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).PluginDisconnected(0xc00068fd40, {0xc000157470, 0x14})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:244 +0x46c
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*client).Disconnect(0xc0000d7ae0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:106 +0x2d9
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).disconnectClient(0xc000714270?, {0xc000157470, 0x14}, {0x3232508, 0xc0000d7ae0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/handler.go:86 +0x55
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).Stop.func1({0xc000157470, 0x14}, {0x3232508, 0xc0000d7ae0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:120 +0x7a
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).visitClients(0xc000374d90, 0xc000763d38)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:186 +0x11f
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1.(*server).Stop(0xc000374d90)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/plugin/v1beta1/server.go:119 +0x73
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).Stop(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:315
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.cleanup(0x322a6f0?, {0x3251ca0, 0xc0006fb480}, 0x0?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager_test.go:335 +0x42
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.TestNewManagerImplStartProbeMode(0x0?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager_test.go:119 +0x126
testing.tRunner(0xc0007081a0, 0x2ef5a08)
	/usr/local/go/src/testing/testing.go:1595 +0x239
```

nothing evident on the git history of the test itself, but it seems @pohly [added it to detect races](https://github.com/kubernetes/kubernetes/commit/fcbb64b93dfb633f880bad9ed1771201843b8e7e), so most probable something in the device manager has changed?

/sig node
/kind flake

### PR DESCRIPTION
**Pull request by aojea**
/kind bug

/kind flake

```release-note
NONE
```

```
$ go test -timeout 30s -run ^TestNewManagerImplStartProbeMode$ k8s.io/kubernetes/pkg/kubelet/cm/devicemanager -v -count 1 -c -race
aojea@aojea:~/src/kubernetes$ stress ./devicemanager.test -test.run TestNewManagerImplStartProbeMode
5s: 162 runs so far, 0 failures
10s: 384 runs so far, 0 failures
15s: 578 runs so far, 0 failures
20s: 783 runs so far, 0 failures
25s: 995 runs so far, 0 failures
30s: 1206 runs so far, 0 failures
35s: 1412 runs so far, 0 failures
40s: 1620 runs so far, 0 failures
45s: 1830 runs so far, 0 failures
50s: 2039 runs so far, 0 failures
55s: 2247 runs so far, 0 failures
1m0s: 2451 runs so far, 0 failures
1m5s: 2668 runs so far, 0 failures
1m10s: 2877 runs so far, 0 failures
1m15s: 3082 runs so far, 0 failures
1m20s: 3287 runs so far, 0 failures
1m25s: 3501 runs so far, 0 failures
1m30s: 3709 runs so far, 0 failures
1m35s: 3915 runs so far, 0 failures
1m40s: 4121 runs so far, 0 failures
1m45s: 4334 runs so far, 0 failures
1m50s: 4542 runs so far, 0 failures
1m55s: 4750 runs so far, 0 failures
2m0s: 4960 runs so far, 0 failures
2m5s: 5170 runs so far, 0 failures
2m10s: 5378 runs so far, 0 failures
```

Fixes: https://github.com/kubernetes/kubernetes/issues/121451

All the credit to @wzshiming , just speeding it up to make the ci signal green again
https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit

I still don't know why it does not show up in  https://storage.googleapis.com/k8s-triage/index.html?pr=1&test=TestNewManagerImplStartProbeMode

### CODE REVIEW

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go [hunk 1]
```diff
@@ -69,8 +69,10 @@ func (c *client) Connect() error {
 		klog.ErrorS(err, "Unable to connect to device plugin client with socket path", "path", c.socket)
 		return err
 	}
+	c.mutex.Lock()
 	c.grpc = conn
 	c.client = client
+	c.mutex.Unlock()
```
Comment (liggitt) [inline pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:75]: do Run() and API() need to RLock to access `c.client`?

**Comment by dims** [comment]
/approve
/lgtm

/hold let's see if we can find a sig-node reviewer quickly else please feel free to remove hold @aojea 

cc @SergeyKanzhelev @tzneal

File: pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:75]: I think the grpc client is thread safe, the race happens with the grpc conn

Comment (liggitt) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:75]: I meant the reassignment to c.client here racing with the access of it. I don't think we need to protect the client methods

Comment (aojea) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:75]: oh, yeah, the problem here seems that there is a Connect() and Disconnect() in parallel, it does not seem Run() or API() are called at the same. time that Connect(), basically those will not work ... but the fact that Disconnect() already had a mutex makes me think those are the only methods that need it

Comment (liggitt) [reply pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go:75]: Ok, this looks like a strict improvement then

/lgtm

**Comment by aojea** [comment]
/assign @ffromani 

since he was reviewing on the other PR

**Comment by wzshiming** [comment]
/lgtm
Thanks

**Comment by liggitt** [comment]
does this need to be picked to any release branches?

**Pull request merged**